test/e2e/gateway/gateway_test.go

```diff
@@ -17,42 +17,95 @@ package gateway
 
 import (
 	"context"
+	"fmt"
 	"testing"
 
 	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gexec"
+	"github.com/projectcontour/contour/pkg/config"
 	"github.com/projectcontour/contour/test/e2e"
 	"github.com/stretchr/testify/require"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	gatewayv1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
 )
 
+var f = e2e.NewFramework(false)
+
 func TestGatewayAPI(t *testing.T) {
+	RegisterFailHandler(Fail)
 	RunSpecs(t, "Gateway API tests")
 }
 
+var _ = BeforeSuite(func() {
+	require.NoError(f.T(), f.Deployment.EnsureResourcesForLocalContour())
+})
+
+var _ = AfterSuite(func() {
+	f.DeleteNamespace(f.Deployment.Namespace.Name, true)
+	gexec.CleanupBuildArtifacts()
+})
+
 var _ = Describe("Gateway API", func() {
-	var f *e2e.Framework
+	var (
+		contourCmd            *gexec.Session
+		contourConfig         *config.Parameters
+		contourConfigFile     string
+		additionalContourArgs []string
+	)
+
+	// Creates specified gateway in namespace and runs namespaced test
+	// body. Modifies contour config to point to gateway.
+	testWithGateway := func(gw *gatewayv1alpha1.Gateway, body e2e.NamespacedTestBody) e2e.NamespacedTestBody {
+		return func(namespace string) {
+			Context(fmt.Sprintf("with gateway %s/%s", namespace, gw.Name), func() {
+				BeforeEach(func() {
+					// Ensure gateway created in this test's namespace.
+					gw.Namespace = namespace
+					// Update contour config to point to specified gateway.
+					contourConfig.GatewayConfig = &config.GatewayParameters{
+						Namespace: namespace,
+						Name:      gw.Name,
+					}
+					require.NoError(f.T(), f.Client.Create(context.TODO(), gw))
+				})
+				AfterEach(func() {
+					require.NoError(f.T(), f.Client.Delete(context.TODO(), gw))
+				})
+
+				body(namespace)
+			})
+		}
+	}
 
 	BeforeEach(func() {
-		f = e2e.NewFramework(GinkgoT())
+		// Contour config file contents, can be modified in nested
+		// BeforeEach.
+		contourConfig = &config.Parameters{}
+
+		// Default contour serve command line arguments can be appended to in
+		// nested BeforeEach.
+		additionalContourArgs = []string{}
+	})
+
+	JustBeforeEach(func() {
+		var err error
+		contourCmd, contourConfigFile, err = f.Deployment.StartLocalContour(contourConfig, additionalContourArgs...)
+		require.NoError(f.T(), err)
+
+		// Wait for Envoy to be healthy.
+		require.NoError(f.T(), f.Deployment.WaitForEnvoyDaemonSetUpdated())
+	})
```

Comment:
why's this have to be in a `JustBeforeEach`, vs. being part of the `BeforeEach`? Perhaps worth a comment.

Reply:
Ah, I think I get this now, but a comment would still be helpful.